### PR TITLE
Adds replace_all_rules and replace_all_synonyms

### PR DESCRIPTION
**Pull request by nunomaduro**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | no     
| Need Doc update   | yes

## Describe your change

Adds methods `replace_all_rules` and `replace_all_synonims`.

### CODE REVIEW

File: algoliasearch/index.py [hunk 1]
```diff
@@ -613,6 +613,24 @@ def batch_synonyms(self, synonyms, forward_to_slaves=False,
...
         return self._req(False, '/synonyms/batch', 'POST', request_options, params, synonyms)
 
+    def replace_all_synonyms(self, synonyms, request_options=None):
+        """
+        Replace all synonyms in the index.
+        @param synonyms the synonyms to upload as a python dictionary.
+               the dictionary must contain an objectID key.
+        """
+        for synonym in synonyms:
+            if 'objectID' not in synonym:
```
Comment (nunomaduro) [inline algoliasearch/index.py]: Make it one line.

File: algoliasearch/index.py [hunk 2]
```diff
@@ -613,6 +613,24 @@ def batch_synonyms(self, synonyms, forward_to_slaves=False,
 
         return self._req(False, '/synonyms/batch', 'POST', request_options, params, synonyms)
 
+    def replace_all_synonyms(self, synonyms, request_options=None):
+        """
+        Replace all synonyms in the index.
+        @param synonyms the synonyms to upload as a python dictionary.
```
Comment (nunomaduro) [inline algoliasearch/index.py]: Replace by a list of python dictionaries.

File: algoliasearch/index.py (hunk 1 above)
Comment (nunomaduro) [reply algoliasearch/index.py]: try accept usage.

File: algoliasearch/index.py [hunk 3]
```diff
@@ -998,6 +1016,24 @@ def save_rule(self, rule, forward_to_replicas=False, request_options=None):
...
         return self._req(False, '/rules/%s' % str(rule['objectID']), 'PUT', request_options, params, rule)
 
+    def replace_all_rules(self, rules, request_options=None):
+        """
+        Replace all rules in the index.
+        @param rules the rules to upload as a python dictionary.
+               the dictionary must contain an objectID key.
+        """
+        for rule in rules:
+            if 'objectID' not in rule:
```
Comment (nunomaduro) [inline algoliasearch/index.py]: try accept usage.

Comment (nunomaduro) [reply algoliasearch/index.py]:
``` try:
            assert not rule['objectID'] == ''
        except (KeyError, AssertionError):
            raise AlgoliaException```

File: algoliasearch/index.py [hunk 4]
```diff
@@ -613,6 +613,24 @@ def batch_synonyms(self, synonyms, forward_to_slaves=False,
 
         return self._req(False, '/synonyms/batch', 'POST', request_options, params, synonyms)
 
+    def replace_all_synonyms(self, synonyms, request_options=None):
+        """
+        Replace all synonyms in the index.
+        @param synonyms the synonyms to upload as a list of python dictionary.
+               the dictionary must contain an objectID key.
+        """
+        for synonym in synonyms:
```
Comment (julienbourdeau) [inline algoliasearch/index.py]: Should we really add this verification in v1's 🤔 
This makes replacesAllSynonyms and `batchSynonyms` behave differently.

File: algoliasearch/index.py [hunk 5]
```diff
@@ -613,6 +613,24 @@ def batch_synonyms(self, synonyms, forward_to_slaves=False,
...
+            if 'objectID' not in synonym:
+                raise AlgoliaException('missing objectID in synonym body')
+            if synonym['objectID'] == '':
+                raise AlgoliaException('objectID in synonym body cannot be empty')
+
+        params = {
+            'replaceExistingSynonyms': True
+        }
+
+        return self._req(False, '/synonyms/batch', 'POST', request_options, params, data=synonyms)
```
Comment (julienbourdeau) [inline algoliasearch/index.py]: Did you not use `batch_synonyms` on purpose because of the requestOptions?

**Pull request merged**